File: simulador/fva_car/remote_api.py
"""Agrupa as leituras e escritas do carro para reduzir chamadas remotas."""

from pathlib import Path


class CarRemoteAPI:
    def __init__(self, sim, robot):
        self.sim = sim
        code = Path(__file__).with_suffix('.lua').read_text(encoding='utf-8')
        self.script = sim.createScript(sim.scripttype_simulation, code)
        sim.setObjectAlias(self.script, 'fva_remote_api')
        sim.setObjectParent(self.script, robot, True)

    def close(self):
        self.sim.removeObjects([self.script])

    def get_state(self, robot):
        return self.sim.callScriptFunction('fvaCarGetState', self.script, robot)

    def set_motors(self, left, right, velocity, force):
        self.sim.callScriptFunction('fvaCarSetMotors', self.script, left, right, velocity, force)

    def set_steering(self, left, right, angle_left, angle_right):
        self.sim.callScriptFunction('fvaCarSetSteering', self.script,
                                    left, right, angle_left, angle_right)

File: simulador/fva_car/remote_api.lua
local sim = require('sim')

function sysCall_init()
end

function fvaCarGetState(robot)
    local linear, angular = sim.getObjectVelocity(robot)
    return {
        time = sim.getSimulationTime(),
        position = sim.getObjectPosition(robot, sim.handle_world),
        quaternion = sim.getObjectQuaternion(robot, sim.handle_world),
        matrix = sim.getObjectMatrix(robot, sim.handle_world),
        linear = linear,
        angular = angular,
    }
end

function fvaCarSetMotors(left, right, velocity, force)
    for _, motor in ipairs({left, right}) do
        sim.setJointTargetVelocity(motor, velocity)
        sim.setJointForce(motor, force)
    end
end

function fvaCarSetSteering(left, right, angleLeft, angleRight)
    sim.setJointTargetPosition(left, angleLeft)
    sim.setJointTargetPosition(right, angleRight)
end
